SIPS/sip-58.md

```diff
@@ -0,0 +1,80 @@
+---
+sip: 58
+title: Emit list of individual fee reclamations and rebates during exchange settlement
+status: Proposed
+author: Jackson Chan (@justinjmoses)
+discussions-to: <https://discord.gg/ShGSzny>
+
+created: 2020-05-11
+---
+
+<!--You can leave these HTML comments in your merged SIP and delete the visible duplicate text guides, they will not appear and may be helpful to refer to if you edit it again. This is the suggested template for new SIPs. Note that an SIP number will be assigned by an editor. When opening a pull request to submit your SIP, please use an abbreviated title in the filename, `sip-draft_title_abbrev.md`. The title should be 44 characters or less.-->
+
+## Simple Summary
+
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the SIP.-->
+
+Emit list of individual fee reclamations and rebates during exchange settlement for Dapps and Synthetix Exchange. Fee reclamation was introduced since [SIP-37](./sip-37.md) but did not emit a breakdown of each individual trade's fee reclamation / rebate amounts.
+
+## Abstract
+
+<!--A short (~200 word) description of the technical issue being addressed.-->
+
+Upgrade the Exchanger `settle()` function to emit individual fee reclaimation / rebate amounts for each trade.
+
+Add an internal function `_settlementsOwing` that will emit an event `ExchangeEntriesSettled` when `Exchanger.settle()` is invoked.
+
+Emit a list for each exchange entry with details about the `amountShouldHaveReceived`, the corresponding `amountReceived`, `reclaimAmount`, `rebateAmount`, and `timestamp` of the entry that is settled in the transaction.
+
+## Motivation
+
+<!--The motivation is critical for SIPs that want to change Synthetix. It should clearly explain why the existing protocol specification is inadequate to address the problem that the SIP solves. SIP submissions without sufficient motivation may be rejected outright.-->
+
+We want to display on Synthetix exchange the corresponding fee reclamation amounts (if any) for each individual exchange made when exchange settlement is invoked.
+
+Currently invoking `settle()` will only emit an event, if any, with the total sum of any fee reclamation or rebate amounts.
+
+## Specification
+
+<!--The technical specification should describe the syntax and semantics of any new feature.-->
+
+### Exchanger.settle() ###
+
+Add an internal function `_settlementsOwing` that will emit an event `ExchangeEntriesSettled` when `Exchanger.settle()` is invoked.
+
+`Exchanger.settle()` will use `_settlementsOwing` to calculate the fee reclamation amounts.
+
+**Event**
+
+Emit an event `ExchangeEntriesSettled` when `Exchanger.settle()` is invoked within `_settlementsOwing`.
+
+```solidity
+event ExchangeEntriesSettled(address indexed from, bytes32 currencyKey, uint numEntries, uint[] amountReceived, uint[] amountShouldHaveReceived, uint[] reclaim, uint[] rebate, uint[] timestamp);
```

Comment:
Why not have an event for each exchange settled rather than aggregate them here? It may be more gas but it will be much easier to cross-reference to the individual exchange. 

Additionally, it should emit as many of the props from the `ExchangeState` as possible to help cross-reference to the individual exchange. So `from`, `src`, `amount`, `dest`, `timestamp` as well as `reclaim` and `rebate`:

How about:

```solidity
ExchangeEntryReclaim(address indexed from, bytes32 src, uint amount, bytes32 dest, uint timestamp, uint reclaim);
ExchangeEntryRebate(address indexed from, bytes32 src, uint amount, bytes32 dest, uint timestamp, uint rebate);
```